File: frontend/src/api/proxy.ts
/**
 * Proxy IP API endpoints
 * Handles node listing, product listing, rental creation and management.
 */

import { apiClient } from './client'

export interface ProxyNode {
  id: number
  ip_address: string
  country: string
  country_code: string
  city: string
  isp: string
  http_port: number
  vless_port: number
  tags: string[]
  status: string
  description: string
}

export interface ProxyProduct {
  id: number
  name: string
  description: string
  duration_days: number
  traffic_limit_gb: number
  price: string
  sort_order: number
}

export interface ProxyCredential {
  http_host: string
  http_port: number
  http_username: string
  http_password: string
  vless_link: string
}

export interface ProxyRental {
  id: number
  node_id: number
  product_id: number
  status: string
  started_at?: string
  expires_at?: string
  traffic_used_bytes: number
  traffic_limit_bytes: number
  created_at: string
  node?: ProxyNode
  product?: ProxyProduct
  credential?: ProxyCredential
}

export interface CreateRentalResponse {
  rental_id: number
  order_id: number
  pay_url?: string
  qr_code?: string
  amount: string
  expires_at: string
}

export interface PaginatedResponse<T> {
  items: T[]
  total: number
  page: number
  page_size: number
  pages: number
}

export async function listNodes(params?: { country_code?: string; tag?: string }): Promise<ProxyNode[]> {
  const { data } = await apiClient.get<ProxyNode[]>('/proxy/nodes', { params })
  return data
}

export async function getNode(id: number): Promise<ProxyNode> {
  const { data } = await apiClient.get<ProxyNode>(`/proxy/nodes/${id}`)
  return data
}

export async function listProducts(): Promise<ProxyProduct[]> {
  const { data } = await apiClient.get<ProxyProduct[]>('/proxy/products')
  return data
}

export async function createRental(nodeId: number, productId: number, payType: string): Promise<CreateRentalResponse> {
  const { data } = await apiClient.post<CreateRentalResponse>('/proxy/rentals', {
    node_id: nodeId,
    product_id: productId,
    pay_type: payType
  })
  return data
}

export async function listRentals(page = 1, pageSize = 20): Promise<PaginatedResponse<ProxyRental>> {
  const { data } = await apiClient.get<PaginatedResponse<ProxyRental>>('/proxy/rentals', {
    params: { page, page_size: pageSize }
  })
  return data
}

export async function getRental(id: number): Promise<ProxyRental> {
  const { data } = await apiClient.get<ProxyRental>(`/proxy/rentals/${id}`)
  return data
}

export async function cancelRental(id: number): Promise<void> {
  await apiClient.post(`/proxy/rentals/${id}/cancel`)
}

export const proxyAPI = {
  listNodes,
  getNode,
  listProducts,
  createRental,
  listRentals,
  getRental,
  cancelRental
}

export default proxyAPI

File: backend/internal/service/proxy.go
package service

import (
	"context"
	"time"

	infraerrors "github.com/meitianwang/fast-frame/internal/pkg/errors"
	"github.com/meitianwang/fast-frame/internal/pkg/pagination"
	"github.com/shopspring/decimal"
)

// --- Sentinel errors ---

var (
	ErrProxyNodeNotFound     = infraerrors.NotFound("PROXY_NODE_NOT_FOUND", "proxy node not found")
	ErrProxyProductNotFound  = infraerrors.NotFound("PROXY_PRODUCT_NOT_FOUND", "proxy product not found")
	ErrProxyRentalNotFound   = infraerrors.NotFound("PROXY_RENTAL_NOT_FOUND", "proxy rental not found")
	ErrProxyNodeNotAvailable    = infraerrors.Conflict("PROXY_NODE_NOT_AVAILABLE", "proxy node is not available for rental")
	ErrProxyRentalNotCancellable = infraerrors.Conflict("RENTAL_NOT_CANCELLABLE", "rental cannot be cancelled in current status")
)

// --- Domain types ---

// ProxyNode represents a VPS-based residential proxy node.
type ProxyNode struct {
	ID           int64
	IPAddress    string
	Country      string
	CountryCode  string
	City         string
	ISP          string
	HTTPPort     int
	VlessPort    int
	VlessNetwork string
	VlessTLS     bool
	VlessSNI     string
	VlessWSPath  string
	Tags         []string
	Status       string
	Description  string
	CreatedAt    time.Time
	UpdatedAt    time.Time
}

// ProxyProduct represents a rental plan.
type ProxyProduct struct {
	ID             int64
	Name           string
	Description    string
	DurationDays   int
	TrafficLimitGB int
	Price          decimal.Decimal
	SortOrder      int
	IsActive       bool
	CreatedAt      time.Time
	UpdatedAt      time.Time
}

// ProxyRental represents a user's active or historical IP rental.
type ProxyRental struct {
	ID                 int64
	UserID             int64
	NodeID             int64
	ProductID          int64
	PaymentOrderID     *int64
	Status             string
	StartedAt          *time.Time
	ExpiresAt          *time.Time
	TrafficUsedBytes   int64
	TrafficLimitBytes  int64
	CreatedAt          time.Time
	UpdatedAt          time.Time

	// Eagerly loaded associations (optional)
	Node    *ProxyNode
	Product *ProxyProduct
}

// ProxyCredential holds generated access credentials for a rental.
type ProxyCredential struct {
	ID           int64
	RentalID     int64
	HTTPUsername string
	HTTPPassword string
	VlessUUID    string
	VlessLink    string
	CreatedAt    time.Time
}

// ProxyTrafficLog records an admin traffic usage update.
type ProxyTrafficLog struct {
	ID          int64
	RentalID    int64
	DeltaBytes  int64
	OperatorID  int64
	Note        string
	CreatedAt   time.Time
}

// --- Repository interfaces ---

type ProxyNodeRepository interface {
	Create(ctx context.Context, node *ProxyNode) error
	GetByID(ctx context.Context, id int64) (*ProxyNode, error)
	Update(ctx context.Context, node *ProxyNode) error
	SoftDelete(ctx context.Context, id int64) error
	List(ctx context.Context, filter ProxyNodeFilter, params pagination.PaginationParams) ([]ProxyNode, *pagination.PaginationResult, error)
	ListAvailable(ctx context.Context, filter ProxyNodeFilter) ([]ProxyNode, error)
	// LockForRental locks the node row in a transaction and returns the current status.
	LockForRental(ctx context.Context, id int64) (*ProxyNode, error)
	SetStatus(ctx context.Context, id int64, status string) error
}

type ProxyProductRepository interface {
	Create(ctx context.Context, p *ProxyProduct) error
	GetByID(ctx context.Context, id int64) (*ProxyProduct, error)
	Update(ctx context.Context, p *ProxyProduct) error
	Delete(ctx context.Context, id int64) error
	List(ctx context.Context, activeOnly bool) ([]ProxyProduct, error)
}

type ProxyRentalRepository interface {
	Create(ctx context.Context, r *ProxyRental) error
	GetByID(ctx context.Context, id int64) (*ProxyRental, error)
	GetByPaymentOrderID(ctx context.Context, orderID int64) (*ProxyRental, error)
	Update(ctx context.Context, r *ProxyRental) error
	List(ctx context.Context, filter ProxyRentalFilter, params pagination.PaginationParams) ([]ProxyRental, *pagination.PaginationResult, error)
	ListByUserID(ctx context.Context, userID int64, params pagination.PaginationParams) ([]ProxyRental, *pagination.PaginationResult, error)
	// ListExpiredActive returns active rentals whose expires_at is past now.
	ListExpiredActive(ctx context.Context, now time.Time, limit int) ([]ProxyRental, error)
}

type ProxyCredentialRepository interface {
	Create(ctx context.Context, cred *ProxyCredential) error
	GetByRentalID(ctx context.Context, rentalID int64) (*ProxyCredential, error)
}

type ProxyTrafficLogRepository interface {
	Create(ctx context.Context, log *ProxyTrafficLog) error
	ListByRentalID(ctx context.Context, rentalID int64) ([]ProxyTrafficLog, error)
}

// --- Filter types ---

type ProxyNodeFilter struct {
	CountryCode string
	Tag         string
	Status      string
}

type ProxyRentalFilter struct {
	UserID  *int64
	NodeID  *int64
	Status  string
}
